pallet/liquidity-pools/src/lib.rs

```diff
@@ -0,0 +1,1065 @@
+
+			let Some(pool) = Pools::<T>::get(id) else {
+				Err(Error::<T>::PoolDoesNotExist)?
+			};
+
+			ensure!(pool.pool_status == PoolStatus::Open, Error::<T>::PoolNotActive);
+
+			ensure!(
+				pool.locked_amount + amount <= pool.max_tokens,
+				Error::<T>::StakingLimitExceeded
+			);
+
+			T::Assets::transfer(pool.asset_id, &who, &vault_account, amount, false)?;
+
+			let user = PoolUsers::<T>::get(id, &who).unwrap_or_default();
+			PoolUsers::<T>::insert(
+				id,
+				&who,
+				UserInfo { amount: user.amount.saturating_add(amount), ..UserInfo::default() },
+			);
+
+			Pools::<T>::try_mutate(id, |pool_info| -> DispatchResult {
+				let pool_info = pool_info.as_mut().ok_or(Error::<T>::PoolDoesNotExist)?;
+				pool_info.locked_amount = pool_info.locked_amount.saturating_add(amount);
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::UserJoined { account_id: who, pool_id: id, amount });
+			Ok(())
+		}
+
+		/// Allows a user to exit from a reward pool, withdrawing their staked tokens.
+		///
+		/// Parameters:
+		/// - `origin`: The account of the user exiting the pool.
+		/// - `id`: The ID of the pool from which the user is exiting.
+		///
+		/// Restrictions:
+		/// - The pool must be in the `Provisioning` status and not closed.
+		/// - The user must have tokens staked in the pool.
+		///
+		/// Emits `UserExited` event if the user successfully exits the pool and claims any rewards.
+		#[pallet::weight(T::WeightInfo::exit_pool())]
+		#[transactional]
+		pub fn exit_pool(origin: OriginFor<T>, id: T::PoolId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let vault_account = Self::get_vault_account(id).unwrap();
+
+			let Some(pool) = Pools::<T>::get(id) else {
+				Err(Error::<T>::PoolDoesNotExist)?
+			};
+
+			ensure!(pool.pool_status == PoolStatus::Open, Error::<T>::PoolNotActive);
+
+			let Some(user_info) = PoolUsers::<T>::get(id, &who) else {
+				Err(Error::<T>::NoTokensStaked)?
+			};
+			ensure!(user_info.amount > Zero::zero(), Error::<T>::NoTokensStaked);
+
+			let amount = user_info.amount;
+			T::Assets::transfer(pool.asset_id, &vault_account, &who, amount, false)?;
+
+			Pools::<T>::try_mutate(id, |pool_info| -> DispatchResult {
+				let pool_info = pool_info.as_mut().ok_or(Error::<T>::PoolDoesNotExist)?;
+				pool_info.locked_amount = pool_info.locked_amount.saturating_sub(amount);
+				Ok(())
+			})?;
+
+			PoolUsers::<T>::remove(id, &who);
+
+			Self::deposit_event(Event::UserExited { account_id: who, pool_id: id, amount });
+			Ok(())
+		}
+
+		/// Claims the reward for a user from a pool that has ended.
+		///
+		/// Users who have staked tokens in a pool that has reached the `Done` status
+		/// can call this function to claim their share of the reward.
+		///
+		/// Parameters:
+		/// - `origin`: The account of the user claiming the reward.
+		/// - `id`: The ID of the pool from which the reward is being claimed.
+		///
+		/// Restrictions:
+		/// - The pool must be in the `Done` status.
```

Comment:
Dont see a `Done` status anymore
```suggestion
		/// - The pool must be in the `Matured` status.
```